src/debug_mode.jl

```diff
@@ -1,5 +1,91 @@
-# TODO: make it non-trivial. See https://github.com/chalk-lab/Mooncake.jl/issues/672
-DebugFRule(rule) = rule
+"""
+    DebugFRule(rule)
+
+Construct a callable equivalent to `rule` but with additional type checking for forward-mode
+AD. Checks:
+- Each `Dual` argument has tangent type matching `tangent_type(typeof(primal))`
+- The returned `Dual` has a correctly-typed tangent
+- Deep structural validation (array sizes, field types, etc.)
+
+Forward-mode counterpart to [`DebugRRule`](@ref).
+
+*Note:* Debug mode significantly slows execution (10-100x) and should only be used for
+diagnosing problems, not production runs.
+
+# Example
+```julia
+using Mooncake
+
+f(x) = sum(x)
+rule = build_frule(zero_dual(f), randn(10); debug_mode=true)
+
+# Valid input - passes
+result = rule(zero_dual(f), Dual(randn(10), randn(10)))
+
+# Invalid input - throws ErrorException wrapping the underlying validation error
+rule(zero_dual(f), Dual(randn(10), randn(11)))  # Size mismatch
```

Comment:
> rule(zero_dual(f), Dual(randn(10), randn(11))) # Size mismatch

I don't think this is correct -- both `frule` and `rrule` only dispatch on argument types, _NOT_ their values.